### PR TITLE
[rpc] refactor GRPC client

### PR DESCRIPTION
**Pull request by zhijunfu**
## Why are these changes needed?

Currently application grpc clients still have some knowledge on GRPC unary implementations. These should be moved into GRPC client for better code abstraction. Also considering that we plan to add support for GRPC stream, which will have a different implementation from GRPC unary, this refactor would also make it easier to add stream support.

This PR also provides a macro to make it easier to add new grpc clients, which is similar to existing `RAY_CORE_WORKER_RPC_HANDLER` used in core worker.

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://ray.readthedocs.io/en/latest/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failure rates at https://ray-travis-tracker.herokuapp.com/.

### CODE REVIEW

**Review by ericl — COMMENTED** [review]
Looks fine to me at a high level. @edoakes any comments?

File: src/ray/rpc/grpc_client.h [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+
+#include "ray/common/grpc_util.h"
+#include "ray/common/status.h"
+#include "ray/rpc/client_call.h"
+
+namespace ray {
+namespace rpc {
+
+// Call a specific rpc method.
+#define RPC_CALL_METHOD(SERVICE, METHOD, request, callback) \
```
Comment (edoakes) [inline src/ray/rpc/grpc_client.h]: Does this assume there is a field defined called "rpc_client_"? That seems a little too magical to me, I'd prefer to just have the user always specify the client (just rename RPC_CALL_METHOD_WITH_CLIENT TO RPC_CALL and remove this one)

Comment (edoakes) [reply src/ray/rpc/grpc_client.h]: Also please improve the comment to specify what this specifically does and why we have it

File: src/ray/rpc/gcs_server/gcs_rpc_client.h [hunk 2]
```diff
@@ -81,10 +73,7 @@ class GcsRpcClient {
   /// \param callback The callback function that handles reply from server.
   void UpdateActorInfo(const UpdateActorInfoRequest &request,
```
Comment (edoakes) [inline src/ray/rpc/gcs_server/gcs_rpc_client.h]: Looks like all of these just call into the macro - is it possible to entirely avoid writing the boilerplate code?

File: src/ray/rpc/object_manager/object_manager_client.h [hunk 3]
```diff
@@ -48,19 +42,17 @@ class ObjectManagerClient {
   /// \param request The request message.
   /// \param callback The callback function that handles reply from server
   void Push(const PushRequest &request, const ClientCallback<PushReply> &callback) {
-    client_call_manager_.CreateCall<ObjectManagerService, PushRequest, PushReply>(
-        *stubs_[push_rr_index_++ % num_connections_],
-        &ObjectManagerService::Stub::PrepareAsyncPush, request, callback);
+    RPC_CALL_METHOD_WITH_CLIENT(ObjectManagerService, Push, request, callback,
+                                rpc_clients_[push_rr_index_++ % num_connections_]);
```
Comment (edoakes) [inline src/ray/rpc/object_manager/object_manager_client.h]: I know it's not introduced by your PR but mind adding a comment about what this magical indexing does?

File: src/ray/rpc/worker/core_worker_client.h [hunk 4]
```diff
@@ -249,21 +224,21 @@ class CoreWorkerClient : public std::enable_shared_from_this<CoreWorkerClient>,
...
-              if (seq_no > max_finished_seq_no_) {
-                max_finished_seq_no_ = seq_no;
-              }
-              rpc_bytes_in_flight_ -= task_size;
-              RAY_CHECK(rpc_bytes_in_flight_ >= 0);
-            }
-            SendRequests();
-            callback(status, reply);
-          });
+      auto rpc_callback = [this, this_ptr, seq_no, task_size, callback](
```
Comment (edoakes) [inline src/ray/rpc/worker/core_worker_client.h]: This might be a good time to move the implementation into a separate .cc file

**Review by edoakes — COMMENTED** [review]
Awesome, definitely an improvement! Just some style/nit comments

File: src/ray/rpc/object_manager/object_manager_client.h (hunk 3 above)
Comment (zhijunfu) [reply src/ray/rpc/object_manager/object_manager_client.h]: sure.

File: src/ray/rpc/grpc_client.h (hunk 1 above)
Comment (zhijunfu) [reply src/ray/rpc/grpc_client.h]: Makes sense. thanks.

File: src/ray/rpc/gcs_server/gcs_rpc_client.h (hunk 2 above)
Comment (zhijunfu) [reply src/ray/rpc/gcs_server/gcs_rpc_client.h]: Done. Please take another look:)

File: src/ray/rpc/worker/core_worker_client.h (hunk 4 above)
Comment (zhijunfu) [reply src/ray/rpc/worker/core_worker_client.h]: sure. I can do this in next PR.

**Review by ericl — APPROVED** [review]

**Pull request merged**